### PR TITLE
Fix osu! (slider) combo not matching expectations when classic mod is disabled

### LINKED ISSUE #21815: PP values displayed in lazer client and on website aren't the same
### Type

Game behaviour

### Bug description

I set a score on Lazer. The client showed me I got 89pp. But the [score page on the website](https://lazer.ppy.sh/scores/2165163328) shows 71pp.
Another more different [play on the website shows 68pp](https://lazer.ppy.sh/scores/2165314696), but in the client was 81pp. (Second screenshot)

### Screenshots or videos

osu_2022-12-27_19-02-12
osu_2022-12-27_18-59-12

### Version

2022.1227.0

### Logs

[database.log](https://github.com/ppy/osu/files/10310005/database.log)
[input.log](https://github.com/ppy/osu/files/10310006/input.log)
[network.log](https://github.com/ppy/osu/files/10310007/network.log)
[performance-audio.log](https://github.com/ppy/osu/files/10310008/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/10310009/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/10310010/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/10310011/performance-update.log)
[performance.log](https://github.com/ppy/osu/files/10310012/performance.log)
[runtime.log](https://github.com/ppy/osu/files/10310013/runtime.log)

### PR DESCRIPTION
**Pull request by peppy**
RFC

- [x] Depends on https://github.com/ppy/osu/pull/25341

See 818432fab4279b6691286483b6be88eb663e84c3's diff for the important part.

Should close https://github.com/ppy/osu/issues/21815. This would also bring the "slider end combo break" behaviour in line with expectations for non-classic mod. Fixing in classic mod is still a mess (see https://github.com/ppy/osu/issues/11769).

There two ways of looking at this:

- This brings things close in line with expectations so it seems like a good starting point to build off.
- This doesn't fix the other issues we're attempting to navigate so it's pointless to even consider.

### CODE REVIEW

File: osu.Game/Rulesets/Scoring/HitResult.cs [hunk 1]
```diff
@@ -350,6 +350,9 @@ public static void ValidateHitResultPair(HitResult maxResult, HitResult minResul
             if (maxResult.IsBonus() && minResult != HitResult.IgnoreMiss)
                 throw new ArgumentOutOfRangeException(nameof(minResult), $"{HitResult.IgnoreMiss} is the only valid minimum result for a {maxResult} judgement.");
 
+            if (minResult == HitResult.IgnoreMiss)
+                return;
```
Comment (bdach) [inline osu.Game/Rulesets/Scoring/HitResult.cs:354]: Should probably update `HitResultTest.TestValidResultPairs()` with coverage of this.

File: osu.Game.Rulesets.Osu/Objects/SliderTailCircle.cs [hunk 2]
```diff
@@ -9,16 +9,28 @@ namespace osu.Game.Rulesets.Osu.Objects
 {
     public class SliderTailCircle : SliderEndCircle
     {
+        /// <summary>
+        /// Whether to treat this <see cref="SliderHeadCircle"/> as a normal <see cref="HitCircle"/> for judgement purposes.
+        /// If <c>false</c>, this <see cref="SliderHeadCircle"/> will be judged as a <see cref="SliderTick"/> instead.
+        /// </summary>
+        public bool ClassicSliderBehaviour;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Objects/SliderTailCircle.cs:16]: Flag should work like the `SliderHeadCircle` flag was made to in bf9f20705f06e3094ae7e772f4ac59fa45472856, i.e. it should only be required to set it on the slider.

In fact:

image

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
> There two ways of looking at this:
> 
>     * This brings things close in line with expectations so it seems like a good starting point to build off.
> 
>     * This doesn't fix the other issues we're attempting to navigate so it's pointless to even consider.

As for this, I think we're no worse-off with this diff than we were previously, so may as well entertain it and see where we end up.

**Comment by smoogipoo** [comment]
I'm not sure whether to do this now or wait to get slider judgements behaving the same as osu-stable first, because I think we can probably agree that the legacy mechanics are something that we have to take forward at this point...

File: osu.Game.Rulesets.Osu.Tests/OsuDifficultyCalculatorTest.cs [hunk 3]
```diff
@@ -15,22 +15,22 @@ public class OsuDifficultyCalculatorTest : DifficultyCalculatorTest
...
+        [TestCase(1.743180218215227d, 54, "zero-length-sliders")]
         public void TestClockRateAdjusted(double expectedStarRating, int expectedMaxCombo, string name)
             => Test(expectedStarRating, expectedMaxCombo, name, new OsuModDoubleTime());
 
-        [TestCase(6.710442985146793d, 239, "diffcalc-test")]
-        [TestCase(0.42506480230838789d, 4, "very-fast-slider")]
-        [TestCase(1.4386882251130073d, 54, "zero-length-sliders")]
+        [TestCase(6.710442985146793d, 272, "diffcalc-test")]
+        [TestCase(0.42506480230838789d, 6, "very-fast-slider")]
+        [TestCase(1.4386882251130073d, 63, "zero-length-sliders")]
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/OsuDifficultyCalculatorTest.cs:33]: Why did these assertions change? I wouldn't expect them to (and tests are failing on this).

This test covers classic mode, so nothing should be changing in that case.

File: osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs [hunk 4]
```diff
@@ -465,11 +465,21 @@ private void assertAllMaxJudgements()
             }, () => Is.EqualTo(judgementResults.Select(j => (j.HitObject, j.Judgement.MaxResult))));
         }
 
-        private bool assertHeadMissTailTracked() => judgementResults[^2].Type == HitResult.SmallTickHit && !judgementResults.First().IsHit;
+        private void assertHeadMissTailTracked()
+        {
+            AddAssert("Tracking retained", () => judgementResults[^2].Type, () => Is.EqualTo(HitResult.LargeTickHit));
+            AddAssert("Slider head misseed", () => judgementResults.First().IsHit, () => Is.False);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs:471]:
```suggestion
            AddAssert("Slider head missed", () => judgementResults.First().IsHit, () => Is.False);
```

File: osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs [hunk 5]
```diff
@@ -18,7 +18,7 @@ public class HitResultTest
         [TestCase(new[] { HitResult.IgnoreHit }, new[] { HitResult.IgnoreMiss, HitResult.ComboBreak })]
         public void TestValidResultPairs(HitResult[] maxResults, HitResult[] minResults)
         {
-            HitResult[] unsupportedResults = HitResultExtensions.ALL_TYPES.Where(t => !minResults.Contains(t)).ToArray();
+            HitResult[] unsupportedResults = HitResultExtensions.ALL_TYPES.Where(t => t != HitResult.IgnoreMiss && !minResults.Contains(t)).ToArray();
```
Comment (bdach) [inline osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs:21]: Hmm I would have expected this test update to look closer to this:

```diff
diff --git a/osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs b/osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs
index 72acd18c5b..e003c9c534 100644
--- a/osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs
+++ b/osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs
@@ -11,14 +11,14 @@ namespace osu.Game.Tests.Rulesets.Scoring
     [TestFixture]
     public class HitResultTest
     {
-        [TestCase(new[] { HitResult.Perfect, HitResult.Great, HitResult.Good, HitResult.Ok, HitResult.Meh }, new[] { HitResult.Miss })]
-        [TestCase(new[] { HitResult.LargeTickHit }, new[] { HitResult.LargeTickMiss })]
-        [TestCase(new[] { HitResult.SmallTickHit }, new[] { HitResult.SmallTickMiss })]
+        [TestCase(new[] { HitResult.Perfect, HitResult.Great, HitResult.Good, HitResult.Ok, HitResult.Meh }, new[] { HitResult.Miss, HitResult.IgnoreMiss })]
+        [TestCase(new[] { HitResult.LargeTickHit }, new[] { HitResult.LargeTickMiss, HitResult.IgnoreMiss })]
+        [TestCase(new[] { HitResult.SmallTickHit }, new[] { HitResult.SmallTickMiss, HitResult.IgnoreMiss })]
         [TestCase(new[] { HitResult.LargeBonus, HitResult.SmallBonus }, new[] { HitResult.IgnoreMiss })]
         [TestCase(new[] { HitResult.IgnoreHit }, new[] { HitResult.IgnoreMiss, HitResult.ComboBreak })]
         public void TestValidResultPairs(HitResult[] maxResults, HitResult[] minResults)
         {
-            HitResult[] unsupportedResults = HitResultExtensions.ALL_TYPES.Where(t => t != HitResult.IgnoreMiss && !minResults.Contains(t)).ToArray();
+            HitResult[] unsupportedResults = HitResultExtensions.ALL_TYPES.Where(t => !minResults.Contains(t)).ToArray();
 
             Assert.Multiple(() =>
             {

```

File: osu.Game.Rulesets.Osu.Tests/OsuDifficultyCalculatorTest.cs (hunk 3 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/OsuDifficultyCalculatorTest.cs:33]: I applied this as a batch change before applying one of your review fixes, which updated to incorrect expectations. Oops.

File: osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs (hunk 5 above)
Comment (peppy) [reply osu.Game.Tests/Rulesets/Scoring/HitResultTest.cs:21]: Looks better. The structure of this test method is very hard for me to parse.

**Review by bdach — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**